2022/Q2/artifacts/pos-model/Staking.tla

```diff
@@ -93,56 +104,65 @@ Init ==
 
 
 \* delegate `amount` coins to Validator
-Delegate(sender, amount) ==
+Delegate(sender, validator, amount) ==
     LET fail ==
         \/ amount < 0
         \/ amount > balanceOf[sender]
     IN
     /\ lastTx' = [ id |-> nextTxId, tag |-> "delegate", fail |-> fail,
-                   sender |-> sender, toAddr |-> Validator, value |-> amount ]
+                   sender |-> sender, toAddr |-> validator, value |-> amount ]
     /\ failed' = (fail \/ failed)
     /\  IF fail
         THEN
-          UNCHANGED <<balanceOf, unbonded, delegated>>
+          UNCHANGED <<balanceOf, unbonded, delegated, bonded>>
         ELSE
           \* transaction succeeds
           \* update the balance of 'sender'
           /\ balanceOf' = [ balanceOf EXCEPT ![sender] = @ - amount]
-          /\ delegated' = [ delegated EXCEPT ![PipelineLength, sender] = @ + amount]
+          /\ delegated' = [ n \in 1..UnbondingLength, user \in UserAddrs, val \in ValidatorAddrs |-> 
+                     IF n >= PipelineLength /\ user = sender /\ val = validator
+                     THEN delegated[n, user, val] + amount
+                     ELSE delegated[n, user, val]
+                    ]
+          /\ bonded' = [ n \in 1..UnbondingLength, user \in ValidatorAddrs |-> 
+                     IF n >= PipelineLength /\ user = sender
+                     THEN bonded[n, user] + amount
+                     ELSE bonded[n, user]
+                    ]
           /\ UNCHANGED unbonded
 
 
-\* unbond `amount` coins from Validator
-Unbond(sender, amount) ==
+\* unbond `amount` tokens from Validator
+Unbond(sender, validator, amount) ==
     LET fail ==
         \/ amount < 0
-        \/ sender = Validator
-        \/ amount > delegated[PipelineLength,sender]
+        \/ amount > delegated[UnbondingLength, sender, validator]
     IN
     /\ lastTx' = [ id |-> nextTxId, tag |-> "unbond", fail |-> fail,
-                   sender |-> sender, toAddr |-> Validator, value |-> amount ]
+                   sender |-> sender, toAddr |-> validator, value |-> amount ]
     /\ failed' = (fail \/ failed)
     /\  IF fail
         THEN
-          UNCHANGED <<balanceOf, unbonded, delegated>>
+          UNCHANGED <<balanceOf, unbonded, delegated, bonded>>
         ELSE
           \* transaction succeeds
           /\ unbonded' = [ unbonded EXCEPT ![UnbondingLength, sender] = @ + amount ]
-          /\ delegated' = [ delegated EXCEPT ![PipelineLength, sender] = @ - amount ]
+          /\ delegated' = [ delegated EXCEPT ![UnbondingLength, sender, validator] = @ - amount ]
+          /\ bonded' = [ bonded EXCEPT ![UnbondingLength, validator] = @ - amount]
           /\ UNCHANGED  balanceOf
+          
 
-\* withdraw unbonded coins
+\* withdraw unbonded tokens
 Withdraw(sender) ==
     LET fail ==
-        \/ sender = Validator
         \/ unbonded[1, sender] <= 0
```

Comment:
[This is just a question of interpretation]
Is this really a failure then the value is 0?
Should it just be a successful withdrawal of 0 tokens? Note that, e.g., in Delegate we allow 0 tokens

Reply:
It is not really. I think it does not matter, but I'll make it consistent.